### PR TITLE
Fix using shared libraries and -fvisibility=hidden

### LINKED ISSUE #6178: `inline static` variables in headers break CUDA kernels of a library loaded with `RTLD_DEEPBIND`
**Describe the bug**

Some variables declared in headers as `inline static` prevent loading dynamic library with the `RTLD_DEEPBIND` flag.

So far I only encountered this bug with the CUDA backend, in my case one dynamic library called `Kokkos::initialize()` while another defined the kernels. Since the variables in [Kokkos::Impl::CudaInternal](https://github.com/kokkos/kokkos/blob/master/core/src/Cuda/Kokkos_Cuda_Instance.hpp#L104-L150) are `inline static`, they are defined in each library defining CUDA kernels as well as in `libkokkoscore.so`.
When loading a library with the `RTLD_DEEPBIND` flag its symbols are not overriden by those in `libkokkoscore.so`, therefore calling `Kokkos::initlialize()` would only initialize the variables in `libkokkoscore.so`, but not those exported by `inline static` and present in the other library.
Kernel launches from the loaded library would then throw an unrelated error.

From a quick `grep -r "inline static .* = .*;$"` in all sources, the HIP backend is also impacted.

**Please also include the following items to support reproducing the bug**
1. compilers: clang-11 (host and CUDA)
2. Kokkos release: 4.0.1
3. platform and backend: linux, CUDA
4. cmake configure command:
```
cmake -S . -B ./build -DKokkos_ROOT=$KOKKOS_ROOT -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DKokkos_ENABLE_CUDA=ON -DKokkos_ENABLE_OPENMP=ON -DKokkos_ARCH_AMPERE80=ON
```
5. output from cmake command

<details>

```
-- The C compiler identification is Clang 11.0.1
-- The CXX compiler identification is Clang 11.0.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found OpenMP_C: -fopenmp=libomp (found version "5.0") 
-- Found OpenMP_CXX: -fopenmp=libomp (found version "5.0") 
-- Found OpenMP: TRUE (found version "5.0")  
-- Enabled Kokkos devices: OPENMP;CUDA
-- kokkos_launch_compiler is enabled globally. C++ compiler commands with -DKOKKOS_DEPENDENCE will be redirected to the appropriate compiler for Kokkos
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    Kokkos_ARCH_AMPERE80
```

</details>

6. code needed to reproduce the bug

<details><summary>`kernels.cpp`</summary>

```c++
#include "Kokkos_Core.hpp"

extern "C" double time_step(size_t N, double dx, const Kokkos::View<double*>& u, const Kokkos::View<double*>& c)
{
    double dt_global = INFINITY;

    Kokkos::parallel_reduce(Kokkos::RangePolicy<>(0, N),
    KOKKOS_LAMBDA(const Kokkos::RangePolicy<>::index_type i, double& dt) {
        double max_cx = Kokkos::max(Kokkos::abs(u[i] + c[i]), Kokkos::abs(u[i] - c[i]));
        dt = Kokkos::min(dt, dx / max_cx);
    }, Kokkos::Min<double>(dt_global));

    return dt_global;
}
```

</details>

<details><summary>`main.cpp`</summary>

```c++
#include "Kokkos_Core.hpp"
#include <dlfcn.h>

double time_step(size_t N, double dx, const Kokkos::View<double*>& u, const Kokkos::View<double*>& c);
using time_step_fptr_t = decltype(time_step)*;

int main()
{
    Kokkos::initialize();

    // Removing 'RTLD_DEEPBIND' will prevent the problem
    void* kernel_lib = dlopen("./build/libkernels.so", RTLD_LAZY | RTLD_LOCAL | RTLD_DEEPBIND);
    if (kernel_lib == nullptr) return 1;
    auto time_step_fptr = (time_step_fptr_t) dlsym(kernel_lib, "time_step");
    if (time_step_fptr == nullptr) return 2;

    {
        size_t N = 10;
        double dx = 0.1;

        Kokkos::View<double*> u("u", N), c("c", N);

        time_step_fptr(N, dx, u, c);
    }

    Kokkos::finalize();
}
```

</details>

<details><summary>`CMakeLists.txt`</summary>

```cmake
cmake_minimum_required(VERSION 3.20)
project(bug_repo_kokkos)

set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_EXTENSIONS OFF)
set(BUILD_SHARED_LIBS ON)

find_package(Kokkos REQUIRED)
    
add_library(kernels SHARED kernels.cpp)
target_link_libraries(kernels PUBLIC Kokkos::kokkos)
    
add_executable(main main.cpp)
target_link_libraries(main PUBLIC Kokkos::kokkos)
```

</details>

7. command line needed to reproduce the bug: `./build/main`
8. please also attach the `KokkosCore_config.h` header file (generated during the build);

<details><summary>`KokkosCore_config.h`</summary>

```c++

#if !defined(KOKKOS_MACROS_HPP) || defined(KOKKOS_CORE_CONFIG_H)
#error \
    "Do not include KokkosCore_config.h directly; include Kokkos_Macros.hpp instead."
#else
#define KOKKOS_CORE_CONFIG_H
#endif

// KOKKOS_VERSION % 100 is the patch level
// KOKKOS_VERSION / 100 % 100 is the minor version
// KOKKOS_VERSION / 10000 is the major version
#define KOKKOS_VERSION 40001
#define KOKKOS_VERSION_MAJOR 4
#define KOKKOS_VERSION_MINOR 0
#define KOKKOS_VERSION_PATCH 1

/* Execution Spaces */
/* #undef KOKKOS_ENABLE_SERIAL */
#define KOKKOS_ENABLE_OPENMP
/* #undef KOKKOS_ENABLE_OPENACC */
/* #undef KOKKOS_ENABLE_OPENMPTARGET */
/* #undef KOKKOS_ENABLE_THREADS */
#define KOKKOS_ENABLE_CUDA
/* #undef KOKKOS_ENABLE_HIP */
/* #undef KOKKOS_ENABLE_HPX */
/* #undef KOKKOS_ENABLE_MEMKIND */
/* #undef KOKKOS_ENABLE_LIBRT */
/* #undef KOKKOS_ENABLE_SYCL */

/* General Settings */
#define KOKKOS_ENABLE_CXX17
/* #undef KOKKOS_ENABLE_CXX20 */
/* #undef KOKKOS_ENABLE_CXX23 */

/* #undef KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE */
/* #undef KOKKOS_ENABLE_CUDA_UVM */
#define KOKKOS_ENABLE_CUDA_LAMBDA
#define KOKKOS_ENABLE_CUDA_CONSTEXPR
/* #undef KOKKOS_ENABLE_IMPL_CUDA_MALLOC_ASYNC */
/* #undef KOKKOS_ENABLE_HIP_RELOCATABLE_DEVICE_CODE */
/* #undef KOKKOS_ENABLE_HIP_MULTIPLE_KERNEL_INSTANTIATIONS */
/* #undef KOKKOS_ENABLE_HPX_ASYNC_DISPATCH */
/* #undef KOKKOS_ENABLE_DEBUG */
/* #undef KOKKOS_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK */
/* #undef KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK */
/* #undef KOKKOS_ENABLE_PROFILING_LOAD_PRINT */
/* #undef KOKKOS_ENABLE_TUNING */
/* #undef KOKKOS_ENABLE_DEPRECATED_CODE_3 */
#define KOKKOS_ENABLE_DEPRECATED_CODE_4
#define KOKKOS_ENABLE_DEPRECATION_WARNINGS
/* #undef KOKKOS_ENABLE_LARGE_MEM_TESTS */
#define KOKKOS_ENABLE_COMPLEX_ALIGN
#define KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
/* #undef KOKKOS_OPT_RANGE_AGGRESSIVE_VECTORIZATION */
/* #undef KOKKOS_ENABLE_AGGRESSIVE_VECTORIZATION */
/* #undef KOKKOS_ENABLE_IMPL_MDSPAN */

/* TPL Settings */
/* #undef KOKKOS_ENABLE_HWLOC */
/* #undef KOKKOS_USE_LIBRT */
/* #undef KOKKOS_ENABLE_HBWSPACE */
#define KOKKOS_ENABLE_LIBDL
/* #undef KOKKOS_ENABLE_LIBQUADMATH */
#define KOKKOS_IMPL_CUDA_CLANG_WORKAROUND

/* #undef KOKKOS_ARCH_SSE42 */
/* #undef KOKKOS_ARCH_ARMV80 */
/* #undef KOKKOS_ARCH_ARMV8_THUNDERX */
/* #undef KOKKOS_ARCH_ARMV81 */
/* #undef KOKKOS_ARCH_ARMV8_THUNDERX2 */
/* #undef KOKKOS_ARCH_A64FX */
/* #undef KOKKOS_ARCH_AMD_AVX2 */
/* #undef KOKKOS_ARCH_AVX */
/* #undef KOKKOS_ARCH_AVX2 */
/* #undef KOKKOS_ARCH_AVX512XEON */
/* #undef KOKKOS_ARCH_KNC */
/* #undef KOKKOS_ARCH_AVX512MIC */
/* #undef KOKKOS_ARCH_POWER7 */
/* #undef KOKKOS_ARCH_POWER8 */
/* #undef KOKKOS_ARCH_POWER9 */
/* #undef KOKKOS_ARCH_INTEL_GEN */
/* #undef KOKKOS_ARCH_INTEL_DG1 */
/* #undef KOKKOS_ARCH_INTEL_GEN9 */
/* #undef KOKKOS_ARCH_INTEL_GEN11 */
/* #undef KOKKOS_ARCH_INTEL_GEN12LP */
/* #undef KOKKOS_ARCH_INTEL_XEHP */
/* #undef KOKKOS_ARCH_INTEL_PVC */
/* #undef KOKKOS_ARCH_INTEL_GPU */
/* #undef KOKKOS_ARCH_KEPLER */
/* #undef KOKKOS_ARCH_KEPLER30 */
/* #undef KOKKOS_ARCH_KEPLER32 */
/* #undef KOKKOS_ARCH_KEPLER35 */
/* #undef KOKKOS_ARCH_KEPLER37 */
/* #undef KOKKOS_ARCH_MAXWELL */
/* #undef KOKKOS_ARCH_MAXWELL50 */
/* #undef KOKKOS_ARCH_MAXWELL52 */
/* #undef KOKKOS_ARCH_MAXWELL53 */
/* #undef KOKKOS_ARCH_PASCAL */
/* #undef KOKKOS_ARCH_PASCAL60 */
/* #undef KOKKOS_ARCH_PASCAL61 */
/* #undef KOKKOS_ARCH_VOLTA */
/* #undef KOKKOS_ARCH_VOLTA70 */
/* #undef KOKKOS_ARCH_VOLTA72 */
/* #undef KOKKOS_ARCH_TURING75 */
#define KOKKOS_ARCH_AMPERE
#define KOKKOS_ARCH_AMPERE80
/* #undef KOKKOS_ARCH_AMPERE86 */
/* #undef KOKKOS_ARCH_ADA89 */
/* #undef KOKKOS_ARCH_HOPPER */
/* #undef KOKKOS_ARCH_HOPPER90 */
/* #undef KOKKOS_ARCH_AMD_ZEN */
/* #undef KOKKOS_ARCH_AMD_ZEN2 */
/* #undef KOKKOS_ARCH_AMD_ZEN3 */
/* #undef KOKKOS_ARCH_VEGA */
/* #undef KOKKOS_ARCH_VEGA906 */
/* #undef KOKKOS_ARCH_VEGA908 */
/* #undef KOKKOS_ARCH_VEGA90A */
/* #undef KOKKOS_ARCH_NAVI */
/* #undef KOKKOS_ARCH_NAVI1030 */
/* #undef KOKKOS_ARCH_NAVI1100 */
```

</details>

**Any additional info**

Launching a kernel would throw this error:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Kokkos::Impl::ParallelReduce< Cuda > requested too much L0 scratch memory
```

Which is caused by the fact that `Kokkos::Impl::m_maxShmemPerBlock` is uninitialized, even though we called `Kokkos::initialize()`.

I would have expected a `"ERROR device not initialized"` error message from [`Kokkos::Impl::CudaInternal::verify_is_initialized`](https://github.com/kokkos/kokkos/blob/7de2b62f79c48905a377590912d9c2a348e70aee/core/src/Cuda/Kokkos_Cuda_Instance.cpp#L320), but this is not the case, since it checks `m_cudaDev`, another `inline static` variable.

We can confirm that the symbol is defined in each library:
```
$ objdump -t ./build/libkernels.so | grep "maxShmem"
00000000002e0818  w    O .bss	0000000000000004              _ZN6Kokkos4Impl12CudaInternal18m_maxShmemPerBlockE

$ objdump -t kokkos_install_dir/libkokkoscore.so | grep "maxShmem"
00000000003a03c0  w    O .bss	0000000000000004              _ZN6Kokkos4Impl12CudaInternal18m_maxShmemPerBlockE
```
Upon loading `libkernels.so` the symbols should be mapped to those in `libkokkoscore.so`, but this is not the case when using `RTLD_DEEPBIND`.
Then when `Kokkos::initialize()` is called, it only initializes `Kokkos::Impl::m_maxShmemPerBlock` in `libkokkoscore.so`, hence the error mentioned above.

By moving the definition of `inline static` variables to `Kokkos_Cuda_Instance.cpp` their symbols are no longer defined in the library:
```
$ objdump -t ./build/libkernels.so | grep "maxShmem"
0000000000000000       O *UND*	0000000000000000              _ZN6Kokkos4Impl12CudaInternal18m_maxShmemPerBlockE
```
Then the kernel launches without any errors. I also tried this fix on a larger application and it did work successfully.

I should emphazise on the fact that loading `libkernels.so` without the `RTLD_DEEPBIND` flag will not cause any problems, therefore this bug is certainly not major but quite difficult to identify from the misleading error message.

In my case I got unlucky since this flag is set by default by Julia's `Libdl.dlopen`.

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/issues/7029. Fixes #6178. Related to https://github.com/kokkos/kokkos/pull/6224. As reported in https://github.com/kokkos/kokkos/pull/5473#issuecomment-1299102792, we run into issues if using `inline static` variables when downstream code compiles with `-fvisibility-hidden`. This pull request makes sure that we test this case (by adding that flag to all the tests when building with shared libraries) and replace the respective variables for `Cuda` and `HIP`.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

File: cmake/kokkos_tribits.cmake [hunk 1]
```diff
@@ -160,6 +160,11 @@ FUNCTION(KOKKOS_ADD_EXECUTABLE_AND_TEST ROOT_NAME)
             )
         ENDIF()
     ENDIF()
+    # We noticed problems with -fvisibility-hidden for inline static variables
+    # if Kokkos was built as shared library.
+    IF(BUILD_SHARED_LIBS AND NOT KOKKOS_CXX_HOST_COMPILER_ID MATCHES MSVC)
+      TARGET_COMPILE_OPTIONS(${PACKAGE_NAME}_${ROOT_NAME} PRIVATE -fvisibility=hidden)
+    ENDIF()
```
Comment (dalg24) [inline cmake/kokkos_tribits.cmake:167]: Why did you decide to do this instead of using `CMAKE_CXX_VISIBILITY_PRESET=hidden` and `CMAKE_VISIBILITY_INLINES_HIDDEN=ON`?

**Review by dalg24 — COMMENTED** [review]
Plz confirm that the Cuda and HIP builds were indeed failing with shared libraries w/o our replacing inline static data members.

File: cmake/kokkos_tribits.cmake (hunk 1 above)
Comment (masterleinad) [reply cmake/kokkos_tribits.cmake]: I wasn't aware of `CMake` attributes using them now.

File: cmake/kokkos_tribits.cmake [hunk 2]
```diff
@@ -160,6 +160,11 @@ FUNCTION(KOKKOS_ADD_EXECUTABLE_AND_TEST ROOT_NAME)
             )
         ENDIF()
     ENDIF()
+    # We noticed problems with -fvisibility-hidden for inline static variables
```
Comment (nmm0) [inline cmake/kokkos_tribits.cmake]: We could do: [CXX_VISIBILITY_PRESET](https://cmake.org/cmake/help/latest/prop_tgt/LANG_VISIBILITY_PRESET.html) it's slighlty more idomatic

Comment (nmm0) [reply cmake/kokkos_tribits.cmake]: Also, the comment here is a bit weird, since it looks like we are enabling hidden visibility by default (which is a good idea, since it's keeps visibility the same as MSVC)

**Review by nmm0 — COMMENTED** [review]
Minor comments

**Comment by masterleinad** [comment]
> Plz confirm that the Cuda and HIP builds were indeed failing with shared libraries w/o our replacing inline static data members.

Yes

**Review by crtrott — APPROVED** [review]

File: cmake/kokkos_tribits.cmake [hunk 3]
```diff
@@ -160,6 +160,12 @@ FUNCTION(KOKKOS_ADD_EXECUTABLE_AND_TEST ROOT_NAME)
             )
         ENDIF()
     ENDIF()
+    # We noticed problems with -fvisibility=hidden for inline static variables
+    # if Kokkos was built as shared library.
+    IF(BUILD_SHARED_LIBS)
+      SET_PROPERTY(TARGET ${PACKAGE_NAME}_${ROOT_NAME} PROPERTY VISIBILITY_INLINES_HIDDEN ON)
+      SET_PROPERTY(TARGET ${PACKAGE_NAME}_${ROOT_NAME} PROPERTY CXX_VISIBILITY_PRESET hidden)
```
Comment (dalg24) [inline cmake/kokkos_tribits.cmake:167]: Why did you chose set target properties rather than set the CMAKE_ prefixed variables globally?

Comment (masterleinad) [reply cmake/kokkos_tribits.cmake:167]: I preferred the per-target property to better control which targets should use it. It needs precisely to be the test targets and none of the implementation files. With CMake variables this feels harder to control.

Comment (nmm0) [reply cmake/kokkos_tribits.cmake:167]: Yeah I think we should avoid setting CMAKE_* variables if possible

**Review by nmm0 — APPROVED** [review]
Looks good to me!

File: cmake/kokkos_tribits.cmake (hunk 3 above)
Comment (dalg24) [reply cmake/kokkos_tribits.cmake:167]: Why are you saying that.  I am talking about passing it in a few CI builds as argument, not setting it in our cmake files.
